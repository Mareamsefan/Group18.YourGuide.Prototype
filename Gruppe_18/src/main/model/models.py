from sqlalchemy import Table, Column, String, Integer, ForeignKey, DATETIME
from sqlalchemy.orm import relationship
from flask_login import UserMixin
import uuid

from Gruppe_18.src.main.database.sql_alchemy import db

tour_account_association = Table(
    'tour_account_association', db.metadata,
    Column('tour_id', String, ForeignKey('Tour.id')),
    Column('account_id', String, ForeignKey('User.id'))
)


class Account(db.Model, UserMixin):
    __tablename__ = "User"
    id = db.Column(db.String, primary_key=True)
    usertype = Column("usertype", String)
    username = Column("username", String)
    password = Column("password", String)
    phoneNumber = Column("phoneNumber", String)
    emailAddress = Column("emailAddress", String)
    tours = relationship("Tour", secondary=tour_account_association, back_populates="participants")

    def get_id(self):
        return str(self.id)

    def __init__(self, usertype, username, password, phoneNumber, emailAddress):
        self.id = str(uuid.uuid4())
        self.usertype = usertype
        self.username = username
        self.password = password
        self.phoneNumber = phoneNumber
        self.emailAddress = emailAddress

    def __repr__(self):
        return f"({self.id}) {self.username} {self.password} {self.phoneNumber} {self.emailAddress}"


class Tour(db.Model):
    # duration in hours
    # cost in dollars
    __tablename__ = "Tour"
    id = Column("id", String, primary_key=True)
    title = Column("title", String)
    date = Column("date", DATETIME)
    destination = Column("destination", String)
    duration = Column("duration", Integer)
    cost = Column("cost", Integer)
    max_travelers = Column("max_travelers", Integer)
    language = Column("language", String)
    pictureURL = Column("pictureURL", String)
    booked = Column("booked", Integer, default=0)
    participants = relationship("Account", secondary=tour_account_association, back_populates="tours")

    def __init__(self, title, date, destination, duration, cost, max_travelers, language, pictureURL):
        self.id = str(uuid.uuid4())
        self.title = title
        self.date = date
        self.destination = destination
        self.duration = duration
        self.cost = cost
        self.language = language
        self.max_travelers = max_travelers
        self.pictureURL = pictureURL

    def __repr__(self):
        return f"({self.id}) {self.title} {self.destination} {self.duration} {self.cost} {self.language} {self.max_travelers} {self.pictureURL} {self.booked}"

    def get_id(self):
        return str(self.id)
