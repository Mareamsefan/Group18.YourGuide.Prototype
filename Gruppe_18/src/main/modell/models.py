from sqlalchemy import Column, String, Integer, DATETIME
import uuid
from Gruppe_18.src.main.database.sql_alchemy import Base


class Account(Base):
    __tablename__ = "User"
    account_id = Column("id", String, primary_key=True)
    username = Column("username", String)
    password = Column("password", String)
    phoneNumber = Column("phoneNumber", String)
    emailAddress = Column("emailAddress", String)

    def __init__(self, username, password, phoneNumber, emailAddress):
        self.account_id = str(uuid.uuid4())
        self.username = username
        self.password = password
        self.phoneNumber = phoneNumber
        self.emailAddress = emailAddress

    def __repr__(self):
        return f"({self.account_id}) {self.username} {self.password} {self.phoneNumber} {self.emailAddress}"

    def get_id(self):
        return self.id


class Tour(Base):
    # duration in hours
    # cost in dollars
    __tablename__ = "Tour"
    tour_id = Column("id", String, primary_key=True)
    title = Column("title", String)
    date = Column("date", DATETIME)
    destination = Column("destination", String)
    duration = Column("duration", Integer)
    cost = Column("cost", Integer)
    max_travelers = Column("max_travelers", Integer)
    language = Column("language", String)
    pictureURL = Column("pictureURL", String)
    booked = Column("booked", Integer, default=0)

    def __init__(self, title, date, destination, duration, cost, max_travelers, language, pictureURL):
        self.tour_id = str(uuid.uuid4())
        self.title = title
        self.date = date
        self.destination = destination
        self.duration = duration
        self.cost = cost
        self.language = language
        self.max_travelers = max_travelers
        self.pictureURL = pictureURL

    def __repr__(self):
        return f"({self.tour_id}) {self.title} {self.destination} {self.duration} {self.cost} {self.language} {self.max_travelers} {self.pictureURL} {self.booked}"

