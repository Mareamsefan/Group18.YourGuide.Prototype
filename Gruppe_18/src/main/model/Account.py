import json
import uuid


class Account:

    def __init__(self, username, password, phoneNumber, emailAddress):
        self.id = str(uuid.uuid4())
        self.username = username
        self.password = password
        self.phoneNumber = phoneNumber
        self.emailAddress = emailAddress



    def get_id(self):
        return self.id
