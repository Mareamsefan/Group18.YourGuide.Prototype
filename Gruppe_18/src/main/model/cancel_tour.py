import datetime

class personTour:
    def __init__(self, tour_name, user_name, reference_number):
        self.tour_name = tour_name
        self.user_name = user_name
        self.reference_number = reference_number
        self.sign_up_time = datetime.datetime.now()
        self.is_cancelled = False

class TourAdmin:
    def __init__(self):
        self.tours = []


    def add_tour(self, tour_name, user_name, reference_number):
        tour = personTour(tour_name, user_name, reference_number)  # Use personTour instead of Tour
        self.tours.append(tour)
        print(f"User '{user_name}' signed up for '{tour_name}' tour with the reference number '{reference_number}'.")
        return True

    def cancel_tour(self, tour_name, user_name, reference_number):
        current_time = datetime.datetime.now()
        for tour in self.tours:
            if tour.tour_name == tour_name and tour.user_name == user_name and not tour.is_cancelled:
                time_difference = current_time - tour.sign_up_time
                if time_difference.total_seconds() <= 24 * 3600:  # 24 hours in seconds
                    return True
                    print(f"User '{user_name}' has canceled the '{tour_name}' tour with the reference number '{reference_number}' within 24 hours. No penalty.")
                else:
                    print(f"User '{user_name}' has canceled the '{tour_name}' tour with the reference number '{reference_number}' after 24 hours. Penalty may apply.")
                    return False
                tour.is_cancelled = True
                self.notify_guide_and_admin(tour)
                self.remove_tour_from_user(tour)
                self.send_cancellation_email(tour)
                return
        print(f"User '{user_name}' is not signed up for '{tour_name}' tour with the reference number '{reference_number}' or has already canceled.")
        return True

    def notify_guide_and_admin(self, tour):
        # Implement notification to tour guide and admin here
        print(f"Notifying guide and admin about the cancellation of '{tour.tour_name}' by '{tour.user_name}' with the reference number '{tour.reference_number}'.")

    def remove_tour_from_user(self, tour):
        # Implement removing the tour from the user's list here
        print(f"Removing '{tour.tour_name}' from '{tour.user_name}' 's tour list with the reference number '{tour.reference_number}'.")

    def send_cancellation_email(self, tour):
        # Implement sending an email about the cancellation here
        print(f"Sending cancellation email to '{tour.user_name}' for '{tour.tour_name}'.")

# Example usage:
tour_admin = TourAdmin()

# User signs up for a tour
tour_admin.add_tour("Tour 1", "Nicole", "DORMMB_18")

# User cancels a tour within 24 hours
tour_admin.cancel_tour("Tour 1", "Nicole", "DORMMB_18")

# User cancels a tour after 24 hours
# tour_admin.cancel_tour("Tour 1", "Nicole", "DORMMB_18")

