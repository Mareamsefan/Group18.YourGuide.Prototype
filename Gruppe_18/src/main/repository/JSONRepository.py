import json


class JSONRepository:

    def save_to_json(self, entity, filename):
        try:
            with open(filename, "r") as json_file:
                filedata = json.load(json_file)
        except FileNotFoundError:
            filedata = []

        duplicate_found = False
        for item in filedata:
            if item == entity.to_dict():
                duplicate_found = True
                break

        if not duplicate_found:
            data = entity.to_dict()
            filedata.append(data)
        with open(filename, "w") as json_file:
            json.dump(filedata, json_file, indent=4)

    def save_to_stream(self, io_stream):
        try:
            filedata = json.load(io_stream)
        except json.JSONDecodeError:
            filedata = []

        duplicate_found = False
        for item in filedata:
            if item == self.to_dict():
                duplicate_found = True
                break

        if not duplicate_found:
            data = self.to_dict()
            filedata.append(data)

        io_stream.truncate(0)
        io_stream.seek(0)
        json.dump(filedata, io_stream, indent=4)

    def to_dict(self, entity):
        entity_dict = {}
        for key, value in entity.__dict__.items():
            entity_dict[key] = value
        return entity_dict

    def read_from_stream(self, stream):
        items = json.load(stream)
        stream.seek(0)
        return items




    # def from_json(self, entity, ):
