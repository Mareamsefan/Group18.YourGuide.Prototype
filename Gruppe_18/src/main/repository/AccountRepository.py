from sqlite3 import IntegrityError

from Gruppe_18.src.main.repository.JSONRepository import JSONRepository
from Gruppe_18.src.main.model.models import Account, Tour, tour_account_association, guide_tour_association
from Gruppe_18.src.main.repository.TourRepository import TourRepository


class AccountRepository(JSONRepository):
    def __init__(self, session):
        self.session = session
        self.tour_repo = TourRepository(session)

    def get_one_specific_account(self, user_id):
        return self.session.query(Account).filter_by(id=user_id).first()

    def delete_account(self, user_id):
        account_to_delete = self.get_one_specific_account(user_id)

        if account_to_delete:
            self.session.delete(account_to_delete)
            self.session.commit()
            return True

        return False

    def upgrade_usertype_to_guide(self, user_id):
        user = self.session.query(Account).filter_by(id=user_id).first()
        user.usertype = "guide"
        self.session.add(user)
        self.session.commit()
        return True

    def create_account(self, user):
        try:
            if any(value is None for value in
                   [user.username, user.password, user.phoneNumber, user.emailAddress]):
                return False
            else:
                account = Account(id=user.id,
                                  usertype=user.usertype,
                                  username=user.username,
                                  password=user.password,
                                  phoneNumber=user.phoneNumber,
                                  emailAddress=user.emailAddress)

                self.session.add(account)
                self.session.commit()
                return True
        except IntegrityError:
            self.session.rollback()
            return False

    def update_account(self, user_id, new_username, new_telephone_number, new_email):
        user = self.get_one_specific_account(user_id)
        user.username = new_username
        user.phoneNumber = new_telephone_number
        user.emailAddress = new_email
        self.session.add(user)
        self.session.commit()
        return True


    def account_register_to_tour(self, tour_id, user_id):
        existing_registration = self.session.query(tour_account_association).filter_by(
            tour_id=tour_id,
            account_id=user_id
        ).first()

        if existing_registration:
            print("You are already registered for that tour.")
        else:
            tour = self.session.query(Tour).filter_by(id=tour_id).first()
            user = self.session.query(Account).filter_by(id=user_id).first()

            if tour is not None and user is not None:
                tour_account_assoc_obj = tour_account_association.insert().values(
                    tour_id=tour_id,
                    account_id=user_id
                )
                self.tour_repo.book_tour(tour)
                self.session.execute(tour_account_assoc_obj)
                self.session.commit()
                return True
            else:
                print("Tur eller bruker ble ikke funnet.")

    def account_cancel_tour(self, tour_id, user_id):
        tour = self.session.query(Tour).filter_by(id=tour_id).first()
        user = self.session.query(Account).filter_by(id=user_id).first()

        if tour is not None and user is not None:
            stmt = tour_account_association.delete().where(
                tour_account_association.c.tour_id == tour_id,
                tour_account_association.c.account_id == user_id
            )
            self.tour_repo.cancel_booked_tour(tour)
            self.session.execute(stmt)
            self.session.commit()
        else:
            print("Tour or user is not found.")



    def account_logged_in(self, status=False):
        logged_in = status
        return logged_in

