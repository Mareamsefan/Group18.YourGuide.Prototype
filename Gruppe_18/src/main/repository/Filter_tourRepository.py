from Gruppe_18.src.main.repository.JSONRepository import JSONRepository

class FilterRepository (JSONRepository):
    def read_from_stream(self, stream):
