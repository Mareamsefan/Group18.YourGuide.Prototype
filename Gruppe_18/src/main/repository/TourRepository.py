from Gruppe_18.src.main.repository.JSONRepository import JSONRepository

list_with_tours =

class TourRepository (JSONRepository):

    def filter_tour_by_location(tour_list, destination):
        filtered_tours = []
        for tour in tour_list:
            if tour['destination'] == destination:
                filtered_tours.append(tour)
        return filtered_tours

    def filter_tour_by_price(tour_list, max_price, min_price):
        filtered_tours = []
        for tour in tour_list:
            if min_price <= tour['cost'] <= max_price:
                filtered_tours.append(tour)
        return filtered_tours