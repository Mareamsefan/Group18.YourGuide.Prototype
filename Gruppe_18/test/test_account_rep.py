import os
import datetime
import uuid

import pytest
from approvaltests import Options, verify
from approvaltests.scrubbers import scrub_all_guids
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from Gruppe_18.src.main.model.models import Account, db, Tour, tour_account_association
from Gruppe_18.src.main.repository.AccountRepository import AccountRepository
from Gruppe_18.src.main.repository.TourRepository import TourRepository

approval_options = Options().with_scrubber(scrub_all_guids)


@pytest.fixture()
def sqlalchemy_session():
    module_path = os.path.dirname(os.path.abspath(__file__))
    database_name = os.path.join(module_path, "Test.db")
    engine = create_engine(f"sqlite:///{database_name}", echo=True)

    session = sessionmaker(bind=engine)()

    db.metadata.create_all(bind=engine)

    yield session

    session.close()
    db.metadata.drop_all(engine)


@pytest.fixture
def account():
    return Account(
        str(uuid.uuid4()),
        "user",
        "username",
        "password",
        "12345678",
        "user_@gmail.com"
    )


@pytest.fixture
def tour_1():
    return Tour(
        str(uuid.uuid4()),
        "Discover Oslo's Charm",
        datetime.date(2024, 8, 5),
        "Oslo, Norway",
        2,
        2500,
        10,
        "Norwegian",
        "https://www.example.com/oslo-city.jpg"
    )


@pytest.fixture
def account_rep(sqlalchemy_session):
    account = AccountRepository(sqlalchemy_session)
    return account


@pytest.fixture
def tour_rep(sqlalchemy_session):
    tour = TourRepository(sqlalchemy_session)
    return tour


# Testing feature 1.1:
@pytest.mark.parametrize("attribute", ["id", "usertype", "username", "password", "phoneNumber", "emailAddress"])
def test_account_creation_and_saving(account_rep, account, sqlalchemy_session, attribute):
    account_rep.create_account(account)
    saved_account_from_db = account_rep.get_one_specific_account(account.id)

    assert getattr(saved_account_from_db, attribute) == getattr(account, attribute)


# Testing feature 1.1.2:
@pytest.mark.parametrize("missing_column", ["username", "password", "phoneNumber", "emailAddress"])
def test_account_missing_information_returns_false(account, account_rep, sqlalchemy_session, missing_column):
    setattr(account, missing_column, None)

    assert account_rep.create_account(account) is False


# Testing feature 1.3.2 and 1.2.1.4:
def test_account_can_be_deleted(account, account_rep):
    account_rep.create_account(account)
    assert account_rep.delete_account(account.id) is True


# Testing feature 1.2.1.1.1, 1.2.1.1.3 and 1.2.1.1.4
def test_account_can_update_their_information(account, account_rep):
    account_rep.create_account(account)
    new_username = "NewName"
    new_phoneNumber = "NewNumber"
    new_emailAddress = "New@gmail.com"
    account_rep.update_account(account.id, new_username, new_phoneNumber, new_emailAddress)
    updated_account = account_rep.get_one_specific_account(account.id)

    verify(repr(updated_account), options=approval_options)


# Testing feature 1.3.1.1:
def test_account_usertype_can_be_upgraded_to_guide(account, account_rep):
    account_rep.create_account(account)
    account_from_db = account_rep.get_one_specific_account(account.id)
    usertype_before_upgrade = account_from_db.usertype

    assert account_rep.upgrade_usertype_to_guide(account_from_db.id) is True

    account_from_db = account_rep.get_one_specific_account(account.id)

    usertype_after_upgrade = account_from_db.usertype
    assert usertype_before_upgrade == "user" and usertype_after_upgrade == "guide"


# Testing Feature 1.8:
def test_account_can_register_a_tour(account_rep, account, sqlalchemy_session, tour_1, tour_rep):
    account_rep.create_account(account)
    tour_rep.create_tour(tour_1)
    account_from_db = account_rep.get_one_specific_account(account.id)
    tour_from_db = tour_rep.get_specific_tour(tour_1.id)
    assert account_rep.account_register_to_tour(tour_from_db.id, account_from_db.id) is True
    registration_row = sqlalchemy_session.query(tour_account_association).filter_by(
        tour_id=tour_from_db.id,
        account_id=account_from_db.id
    ).first()
    assert registration_row is not None
