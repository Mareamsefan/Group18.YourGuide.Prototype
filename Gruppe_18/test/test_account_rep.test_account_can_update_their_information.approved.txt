(<guid_0>) user NewName password NewNumber New@gmail.com
